### PR TITLE
support deepseekv4 flash fsdp2 +ep training 

### PR DESCRIPTION
**Pull request by meichangsu1**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [x] More Models or Datasets Support

# PR information

This PR adds DeepSeek V4 support across template processing, Transformers model initialization, and training cookbook examples.

1. DeepSeek V4 template support
- Add `DeepseekV4Template` and register it in `twinkle.template`.
- Introduce a dedicated DeepSeek V4 tokenizer wrapper that overrides `apply_chat_template`.
- Add DeepSeek V4 message encoding logic, refered to [https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash/blob/main/encoding/encoding_dsv4.py](url)

2. Reduce Transformers v5 model loading memory pressure
- Refine pretrained model initialization for distributed/FSDP training.
- Add rank-aware empty-model initialization for non-rank0 workers when memory-efficient init is enabled.
- Build empty models from config under `init_empty_weights`, then rely on later FSDP wrapping / broadcast-based loading flow.
- Preserve normal `from_pretrained` behavior for non-memory-efficient paths.
- This is aimed at avoiding excessive peak memory use and improving compatibility with newer Transformers loading behavior.

3. Add DeepSeek V4 FSDP2 + LoRA cookbook
- Add `cookbook/transformers/deepseek_v4_flash.py`.
- Provide an end-to-end local training example with:
  - `DeepseekV4Template`
  - FSDP2/native FSDP strategy
  - memory-efficient initialization
- Current cookbook scope is FSDP2 + LoRA training only. EP is not supported in this example yet.

4. Training correctness and current limitations
- Training accuracy has been aligned with the expected baseline.
- Add dtype alignment handling around LoRA/FSDP integration.
- For NPU, add a temporary workaround to align parameter dtype more aggressively with the base model dtype to keep training behavior correct.
- NPU correctness is aligned, but runtime performance still needs further optimization.

## Experiment results

- DeepSeek V4 FSDP2 + LoRA training flow has been validated.
- Accuracy behavior has been aligned with the expected baseline.
- EP training is not covered by the current cookbook.
- NPU training correctness is available, while performance optimization remains follow-up work.

Below are the two-layer loss alignment results on NPU and GPU. The loss curves overlap almost perfectly, and the loss difference gradually converges to zero in the final stage.

a0412c9124febad049676c9d457954f6
cooperate with @0hujun

### CODE REVIEW

**Comment by tpx818** [comment]
EP support will be added ASAP

File: src/twinkle/model/transformers/transformers.py [hunk 1]
```diff
@@ -42,10 +42,21 @@
...
 from twinkle.utils.framework import Torch
 from twinkle.utils.grad_clip import normalize_and_clip_grad_norm
+from twinkle.utils.transformers_utils import filter_from_config_kwargs
 
 logger = get_logger()
 
 
+def _clone_state_dict_to_cpu(state_dict: Dict[str, Any]) -> Dict[str, Any]:
+    cloned = {}
+    for key, value in state_dict.items():
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这个可以放到其他文件中，比如torch_utils

**Review by tastelikefeet — APPROVED** [review]

File: src/twinkle/model/transformers/transformers.py (hunk 1 above)
Comment (meichangsu1) [reply src/twinkle/model/transformers/transformers.py]: 好的
